과제 제출

steve7867 opened the pull request:
리뷰 부탁드립니다.

커밋 가이드도 제대로 못 지키고 부족한 게 많네요....

steve7867 committed:
과제 제출

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,144 @@
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+
+import java.io.*;
```
코드숨 과정을 진행하는 동안에는 `import *`의 사용을 지양하도록 합시다. 어떤 것을 import 하는지 분명히 명시해두도록 해봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        String method = exchange.getRequestMethod();
+        URI uri = exchange.getRequestURI();
```
가능한 한 재할당을 방지해 주도록 합시다.
```suggestion
        final String method = exchange.getRequestMethod();
        final URI uri = exchange.getRequestURI();
```

steve7867 replied:
궁금한 점이 있습니다.
메소드의 파라미터도 메소드 내에서 재할당 되지 않는다면 final keyword를 붙히는 것이 바람직한가요?
코드의 안정성과 컴파일러 최적화 관점에서 final keyword를 붙혀 재할당을 방지하는 것은 바람직한 코딩이라고 생각합니다만 재할당 되지 않는 모든 파라미터에 final을 붙히게 되면 코드가 장황해 보이지 않을까하는 생각도 듭니다.
실무에서는 어떤 방식으로 코딩을 하는지 궁금합니다.

johngrib replied:
실무에서는 - 이건 사람마다 다 달라서 뭐라고 할 수 없어요. 어떤 사람은 안 붙이고 어떤 사람은 붙입니다. 그래서 정답이 없죠. 이건 자신만의 주관을 갖는 문제이기도 해요. 저는 가능한 한 final을 붙이는 방식으로 작업합니다. 하지만 이렇게 생각하지 않던 시절도 있었어요. @steve7867 님도 아마 계속 변하거나 아니면 하나의 스타일에 정착하시게 될 겁니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,144 @@
+        String content = null;
+        OutputStream outputStream = exchange.getResponseBody();
+
+        if (pathArr.length == 3 && method.equals("GET")) {
+            Long id = Long.valueOf(pathArr[2]);
+            content = findTaskById(id, exchange);
+        }
+        if (pathArr.length == 3 && (method.equals("PUT") || method.equals("PATCH"))) {
+            Long id = Long.valueOf(pathArr[2]);
```
GET, PUT, PATCH, DELETE 같은 HTTP method는 프로토콜에서 결정된 값들이니 enum으로 정의해 보도록 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,28 @@
+        return id;
+    }
+
+    public void setId() {
+        this.id = sequence++;
+    }
```
이 메소드는 스레드 안전할까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,144 @@
+        return content;
+    }
+
+    private String findTaskById(Long id, HttpExchange exchange) throws IOException {
+        String content;
+        Task task = taskMap.get(id);
+
+        if (task == null) {
+            content = "Not Found!";
+            exchange.sendResponseHeaders(404, content.getBytes().length);
+        } else {
+            content = taskToJson(task);
+            exchange.sendResponseHeaders(200, content.getBytes().length);
+        }
+
+        return content;
+    }
```
early return 방식에 대해 조사해 보시고, early return 스타일로 이 메소드를 수정해 보세요.

johngrib reviewed (commented):
반갑습니다 :-)

steve7867 committed:
2번째 제출

steve7867 committed:
Refactor - Delete setResponseHeader()

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/195092332-8ae8a7f5-cad9-4880-a23e-2156600ff707.png)

커밋 메시지를 작성할 때에는 작업의 내용을 커밋 메시지로 작성해 보세요. "2번째 제출"은 저와 준수님 사이에서는 의미있는 문장일 수 있겠지만, 코드 관점에서는 중요하지 않습니다. 작업을 다양하게 쪼개는 연습을 해보세요. 쪼갠 작업마다 작업의 이름을 붙일 수 있을 겁니다. 그 이름을 그대로 커밋 메시지로 작성해서 올려보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/Method.java:
```diff
@@ -0,0 +1,12 @@
+package com.codesoom.assignment;
+
+public enum Method {
```
그냥 메소드는 너무 넓은 의미를 갖고 있습니다. 이렇게 해보시면 어떨까요?
```suggestion
public enum HttpMethod {
```

steve7867 replied:
수정해 보겠습니다. 적절한 이름 감사합니다~

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,165 @@
+        final URI uri = exchange.getRequestURI();
+        final String path = uri.getPath();
+
+        final String[] pathArr = path.split("/");
```
`path`가 null 인 경우를 (IOException가 던져지도록 하지 않고) 직접 처리해 보세요.

steve7867 replied:
path가 null인 경우 NullPointerException이 일어나는 걸로 알고 있는데, 어떤 경우에 IOException이 일어나나요?

johngrib replied:
아 제가 잘못 적었습니다. `(Exception이 던져지도록 하지 않고) 직접 처리해 보세요`가 제가 의도한 문장입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,165 @@
+        if (pathArr.length == 3) {
+            final Long id = Long.valueOf(pathArr[2]);
+
+            if (method.equals(Method.GET))
```
이렇게 하면 `method`가 `null`인 경우에도 NPE가 던져지지 않을 것입니다.
```suggestion
            if (Method.GET.equals(method))
```

steve7867 replied:
전혀 생각하지 못한 부분이네요! 수정하겠습니다.

johngrib replied:
Java 코딩시 흔하게 사용하는 기법이니 꼭 기억해 두세요. 앞으로도 자주 사용하게 되실 거에요.

steve7867 replied:
항상 유념하면서 코딩해야겠습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,165 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public class MyHandler implements HttpHandler {
```
이 클래스에서 다른 클래스로 분리할만한 기능들이 무엇이 있을지 고민해보시고 작업도 해 보세요. 이 파일의 크기를 50줄 정도로 줄여 보세요.

steve7867 replied:
Request를 처리하는 책임을 HttpRequestService에 넘겼습니다.
HttpRequestService 인터페이스를 생성하고 이를 구현한 GetService, PostService, EditService, DeleteService를 생성했습니다.
Task를 추가하고 조회하고 삭제하는 역할을 하는 TaskRepository 클래스를 생성했습니다.
50줄까지는 줄이지 못했고 63줄까지 줄였습니다.😭

johngrib replied:
50줄은 대략적으로 말해본 것이니 꼭 수를 맞추지 않으셔도 됩니다. 수고하셨습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public void setId() {
+        this.id = atomicLong.getAndIncrement();
```
아이디를 생성하는 책임을 이제 다른 파일로 분리해 봅시다. `IdGenerator`라는 클래스를 만들어 보세요.

steve7867 replied:
IdGenerator 클래스를 만들어 Id 생성하는 책임을 여기로 위임했습니다.

steve7867 commented:
> ![image](https://user-images.githubusercontent.com/1855714/195092332-8ae8a7f5-cad9-4880-a23e-2156600ff707.png)
> 
> 커밋 메시지를 작성할 때에는 작업의 내용을 커밋 메시지로 작성해 보세요. "2번째 제출"은 저와 준수님 사이에서는 의미있는 문장일 수 있겠지만, 코드 관점에서는 중요하지 않습니다. 작업을 다양하게 쪼개는 연습을 해보세요. 쪼갠 작업마다 작업의 이름을 붙일 수 있을 겁니다. 그 이름을 그대로 커밋 메시지로 작성해서 올려보세요.

작업 단위로 커밋 작성하는 습관을 들이도록 하겠습니다.

steve7867 committed:
Refactor - Change Method -> HttpMethod

Method enum의 이름을 HttpMethod로 변경.
해당 타입의 method 변수명을 httpMethod로 변경.

steve7867 committed:
Docs - Add documentation for HTTP method

steve7867 committed:
Fix - Prevent NullPointerException

Prevent NPE when httpMethod is null

steve7867 committed:
Fix - Prevent NPE when path is null

path가 null인 경우 400(Bad Request) 응답을 보내고 종료

steve7867 committed:
Refactor - Add Validator class

Validator 클래스를 추가하여 path와 httpMethod가 null인지 검증

steve7867 committed:
Feat - Add HttpStatusCode enum

steve7867 committed:
Fix - Add CREATED(201) to HttpStatusCode enum

steve7867 committed:
Refactor - Http status code number -> HttpStatusCode enum value

steve7867 committed:
Refactor: Add line breaking for readability

steve7867 committed:
Refactor: change setId() -> allocateId()

steve7867 committed:
Refactor: Change setId() -> allocateId()

steve7867 committed:
Feat: Add TaskRepository class

steve7867 committed:
Feat: Create JsonConverter class

Json 형식의 문자열을 Task 객체로, Task 객체를 Json 형식의 문자열로
변환해주는 class

steve7867 committed:
Refactor: Change instance methods to static methods

steve7867 committed:
Refactor: Change private methods to public

steve7867 committed:
Fix: Change code field to public

steve7867 committed:
Feat: Add GetService class

steve7867 committed:
Feat: Create PostService class

steve7867 committed:
Fix: Remove imported class unnecessary

steve7867 committed:
Feat: Create EditService class

steve7867 committed:
Feat: Create DeleteService class

steve7867 committed:
Refactor: Move TaskRepository class into repository/

steve7867 committed:
Fix: Add parameter(Long id)

steve7867 committed:
Feat: Add HttpRequestService interface

HttpRequestService 인터페이스를 생성하고 GetService, PostService,
EditService, DeleteService가 이를 구현

steve7867 committed:
Refactor MyHandler class

request를 처리하는 책임을 HttpRequestService에 위임.

steve7867 committed:
Move JsonConverter, Validator into utils/

steve7867 committed:
Feat: Add editTaskById()

steve7867 committed:
Refactor EditService class

TaskRepository의 editTaskById를 사용

steve7867 committed:
Feat: Add IdGenerator class

steve7867 committed:
Fix: Change IdGenerator class singleton

steve7867 committed:
Refactor Task, TaskRepository class

id 생성 기능을 IdGenerator class에 위임

steve7867 committed:
Feat: Add HttpResponse class

steve7867 committed:
Feat: Add getter

steve7867 committed:
Refactor HttpRequestService and implemented class

Http request를 처리한 후 HttpResponse 타입으로 반환

steve7867 committed:
Refactor: Change variable name

content -> emptyString

johngrib reviewed (commented):
이번에는 커밋의 수가 너무 많습니다! ㅎㅎㅎㅎㅎ
간단한 팁을 드릴게요.

1. 해야 할 작업들을 먼저 종이에 순서대로 적어봅니다.
2. 작업의 수가 너무 많거나 적지 않도록 일을 쪼개거나 적당히 합칩니다.
3. 작업 하나가 파일 3개 이상을 변경한다면 2개로 쪼개도록 합니다.
4. 작업 하나가 5분 안에 끝나거나 너무 미세하다면 다른 작업이랑 합칩니다.
5. 이렇게 정리된 작업들을 순서대로 해나갑니다.
6. 작업 하나를 마칠 때마다 작업 이름을 커밋 메시지로 적습니다.

일단 지금은 커밋 메시지를 한국어로 적어 보도록 합시다. 한국어로 잘 적을 수 있게 되면 그 다음에 영어로 적는 연습을 해보기로 해요.

johngrib commented on app/src/main/java/com/codesoom/assignment/HttpStatusCode.java:
```diff
@@ -0,0 +1,15 @@
+package com.codesoom.assignment;
+
+public enum HttpStatusCode {
+    OK(200),
+    CREATED(201),
+    NO_CONTENT(204),
+    BAD_REQUEST(400),
+    NOT_FOUND(404);
```
좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,65 @@
+    public void handle(HttpExchange exchange) throws IOException {
+        final HttpMethod httpMethod = HttpMethod.valueOf(exchange.getRequestMethod());
+        final URI uri = exchange.getRequestURI();
+        final String path = uri.getPath();
+        final OutputStream outputStream = exchange.getResponseBody();
+
+        if (!Validator.isValid(path, httpMethod)) {
+            exchange.sendResponseHeaders(HttpStatusCode.BAD_REQUEST.code, 0);
+            writeAndFlushAndClose("", outputStream);
+            return;
+        }
```
`uri`를 내부에 갖고 있는 `Path` 클래스를 정의해 보세요. 그 클래스가 `isValid` 메소드를 갖도록 작업할 수 있겠죠.

steve7867 replied:
Path 클래스를 만들고 path와 관련된 작업은 이 클래스로 다 위임했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,65 @@
+        final String[] pathArr = path.split("/");
+
+        Long id = null;
+        if (pathArr.length == 3)
+            id = Long.valueOf(pathArr[2]);
```
if 문에서 절대로 중괄호를 생략하지 않도록 합시다.

steve7867 replied:
![화면 캡처 2022-10-12 231548](https://user-images.githubusercontent.com/115217247/195367567-6626be35-b88d-471d-b220-128631b22906.png)
Intellij 설정을 바꿔서 강제하게끔 했습니다.

steve7867 replied:
한 줄 짜리 if문의 중괄호를 생략하면 어떤 문제가 있는지 여쭤봐도 될까요?
Google Java Style Guide, Java Code Conventions(Oracle)에도 한 줄 짜리 if문에 중괄호를 넣어야 한다고 이야기는 하는데 이유에 대해서는 나와 있지 않아서요.

johngrib replied:
이렇게 하는 이유는 수많은 회사들에서 중괄호를 습관적으로 생략했다가 어쩌다 한 번 날려버린 품질의 비용이 어마어마하기 때문입니다.

Apple의 SSL/TLS 프로토콜에서 발견된 중괄호 생략 때문에 생긴 문제가 제일 유명한 케이스입니다. 중괄호 생략은 아예 생각도 하지 않는 것이 좋습니다. 한 번 실수로 많은 것을 날려먹을 수 있습니다. 그에 비하면 중괄호 넣어주는 건 아주 싸죠. 가성비가 너무 좋은 습관입니다.

https://embeddedgurus.com/barr-code/2014/03/apples-gotofail-ssl-security-bug-was-easily-preventable/

![image](https://user-images.githubusercontent.com/1855714/195602644-e7f5b05f-32ba-4775-a5c0-b8a39f2ef3fb.png)

어떻게 하더라도 무조건 goto fail 로 가는 코드가 우리에게는 잘 보입니다. 하지만 이 작업을 수행했던 뛰어난 프로그래머는 문제가 없다고 생각했었겠죠.

steve7867 replied:
사소한 코딩 습관이 이렇게 중대한 문제로 이어진 케이스는 처음 봅니다.
좋은 코딩 습관을 들이는 게 정말 중요하다는 생각이 드네요.

johngrib replied:
물론 개인 프로젝트 같은 거라면 괜찮습니다 ㅎㅎㅎ 하지만 연봉과 커리어가 걸려 있는 회사 코드에서는 이런 보험과 같은 안전한 습관들이 굉장히 중요합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,65 @@
+        outputStream.close();
+    }
+
+    private HttpRequestService resolveHttpRequestService(HttpMethod httpMethod) {
+        switch (httpMethod) {
+            case GET:
+                return GetService.getInstance();
+            case POST:
+                return PostService.getInstance();
+            case PUT:
+            case PATCH:
+                return EditService.getInstance();
+            case DELETE:
+                return DeleteService.getInstance();
+            default:
+                throw new IllegalArgumentException();
+        }
```
보기 좋습니다. 깔끔하군요!
여기에 추가로 `IllegalArgumentException`이 아니라 이 코드 맥락에 맞춰 따로 세밀하게 정의한 이름을 갖는 예외를 사용하는 것도 좋겠습니다.

steve7867 replied:
네. 상황에 맞는 Exception을 새로 생성했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/HttpResponse.java:
```diff
@@ -0,0 +1,21 @@
+import com.codesoom.assignment.HttpStatusCode;
+
+public class HttpResponse {
+    String content;
+    HttpStatusCode httpStatusCode;
```
```suggestion
    private final String content;
    private final HttpStatusCode httpStatusCode;
```

steve7867 replied:
private final 붙히는 걸 빠뜨렸네요. 
수정하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/services/DeleteService.java:
```diff
@@ -0,0 +1,34 @@
+        return instance;
+    }
+
+    public HttpResponse serviceRequest(Long id, HttpExchange exchange) throws IOException {
```
추가 과제입니다. HTTP 개념이 서비스 레이어까지 침투하지 않도록 수정해 보세요.
조금 어려울 수도 있는데요, 잘 하고 계셔서 추가 과제를 드리고 싶습니다.

steve7867 replied:
HttpRequest 클래스를 만들고 여기서 requestBody를 추출하여 HttpRequestService로 넘겼습니다.

steve7867 commented:
> 이번에는 커밋의 수가 너무 많습니다! ㅎㅎㅎㅎㅎ 간단한 팁을 드릴게요.
> 
> 1. 해야 할 작업들을 먼저 종이에 순서대로 적어봅니다.
> 2. 작업의 수가 너무 많거나 적지 않도록 일을 쪼개거나 적당히 합칩니다.
> 3. 작업 하나가 파일 3개 이상을 변경한다면 2개로 쪼개도록 합니다.
> 4. 작업 하나가 5분 안에 끝나거나 너무 미세하다면 다른 작업이랑 합칩니다.
> 5. 이렇게 정리된 작업들을 순서대로 해나갑니다.
> 6. 작업 하나를 마칠 때마다 작업 이름을 커밋 메시지로 적습니다.
> 
> 일단 지금은 커밋 메시지를 한국어로 적어 보도록 합시다. 한국어로 잘 적을 수 있게 되면 그 다음에 영어로 적는 연습을 해보기로 해요.

팁 감사합니다~
주신 팁대로 커밋하겠습니다.

steve7867 committed:
if문 중괄호 추가, HttpResponse 클래스 필드 접근자 수정 및 재할당 방지

steve7867 committed:
Path.class 추가, Validator.class 삭제

steve7867 committed:
serviceRequest() 파라미터 변경

(Long id, HttpExchange exchange) -> (Long id, String requestBody)

steve7867 committed:
HttpRequest 관련 Exception 추가

steve7867 committed:
HttpRequest.class 추가, Path.class 변경

steve7867 committed:
MyHandler.java 리팩토링

Http request 정보를 추출하는 책임을 HttpRequest에 위임.

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/195603214-55a799af-ef2a-47b0-924d-ee9bf8a484a4.png)

좋습니다! 커밋 메시지를 읽고 어떤 작업이 진행됐는지 파악하기 훨씬 좋아졌습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHandler.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MyHandler implements HttpHandler {
```
기능을 고려해 보면 ServiceDispatcher나 ServiceContainer, ServiceSelector 같은 이름도 괜찮아 보입니다.

steve7867 replied:
그렇네요. ServiceDispatcher라는 이름으로 수정하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/HttpRequest.java:
```diff
@@ -0,0 +1,56 @@
+            String methodName = exchange.getRequestMethod();
+            return HttpMethod.valueOf(methodName);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalHttpRequestMethodException("옳지 않은 Http Method 입니다. GET, POST, PUT, PATCH, DELETE 중에서 선택해 주세요.");
```
예외 메시지를 잘 작성하셨습니다. 좀 투박해 보이는 문장이긴 하지만 기본을 잘 지키셨고, 잘못된 요청을 보낸 사람에게 선택지 목록을 주는 것도 바람직합니다.

steve7867 replied:
감사합니다~😊

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Path.java:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        final String idPart = pathArr[2];
+        try {
+            Long.valueOf(idPart);
+        } catch (NumberFormatException e) {
+            return false;
+        }
```
추가 과제입니다. try / catch 없이 작동하도록 수정해 보세요.

steve7867 replied:
StringValidator 클래스를 따로 생성하고 여기에 문자열이 숫자로만 이루어졌는지 판단하는 메소드를 구현하였습니다.

johngrib reviewed (commented):
주의 깊게 작업을 잘 해나가고 계신 것 같습니다 :-)

steve7867 committed:
MyHandler.java -> ServiceDispatcher.java

steve7867 committed:
StringValidator.java 추가, resolveId와 isValid 수정

resolveId: pathArr 재할당 방지
isValid: idPart 검증 부분 리팩토링

steve7867 committed:
HttpRequestValidator.java 추가 및 HttpRequest.java, Path.java 리팩토링

steve7867 committed:
IdGenerator 상속 금지, IllegalHttpRequestBodyException.java 추가

steve7867 committed:
HttpRequestValidator 기존 메소드 변경 및 기능 추가

기존 메소드: 파라미터의 valid 여부를 리턴하는 방식에서 Exception을 던지는 방식으로 변경.
기능 추가: Http Request Method에 맞는 id와 request body가 존재하는지 검증.

steve7867 committed:
HttpMethod getter 추가 및 HttpRequest, Path 리팩토링

steve7867 committed:
HttpRequestValidator 버그 수정

steve7867 commented:
Request Body가 Json 형식인지 아닌지 검증하는 기능을 추가하려고 하는데 Jackson 라이브러리에는 해당 기능이 없습니다.
만약 다른 라이브러리에 해당 기능이 있다면 다른 라이브러리를 사용해도 괜찮나요?

johngrib commented:
> Request Body가 Json 형식인지 아닌지 검증하는 기능을 추가하려고 하는데 Jackson 라이브러리에는 해당 기능이 없습니다. 만약 다른 라이브러리에 해당 기능이 있다면 다른 라이브러리를 사용해도 괜찮나요?

물론입니다. 다양한 시도를 해보세요. 다른 라이브러리를 사용하셔도 좋고, 만약 욕심이 있으시다면 직접 구현하시는 것도 추천드립니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringValidator.java:
```diff
@@ -0,0 +1,14 @@
+
+public final class StringValidator {
+
+    public static boolean isNumber(String s) {
```
`isNumberFormat`이라는 이름도 괜찮을 것 같네요.

steve7867 replied:
맞습니다.
isNumberFormat의 의미가 더 명확해 보입니다.
수정하겠습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/HttpRequestValidator.java:
```diff
@@ -0,0 +1,41 @@
+
+        final String idPart = pathArr[2];
+        return StringValidator.isNumber(idPart);
+    }
```
이 메소드에 대한 테스트코드를 만들어 보는 것도 좋은 학습이 될 것 같습니다.

steve7867 replied:
테스트 코드를 작성했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/IdGenerator.java:
```diff
@@ -2,7 +2,7 @@
 import java.util.concurrent.atomic.AtomicLong;
 
-public class IdGenerator {
+public final class IdGenerator {
```
좋습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/ServiceDispatcher.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ServiceDispatcher implements HttpHandler {
```
JavaDoc 작성법을 조사해 보신 다음, 이 클래스의 역할과 책임을 설명하는 JavaDoc을 작성해 보세요.

steve7867 replied:
JavaDoc 작성하는 게 코딩하는 것보다 더 어렵네요.
이것도 많은 연습이 필요할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/IllegalHttpRequestException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesoom.assignment.exceptions;
+
+public class IllegalHttpRequestException extends Exception{
```
이 정도로도 충분하겠죠.
```suggestion
public class IllegalHttpRequestException extends RuntimeException {
```

steve7867 replied:
`RuntimeException` 상속을 제안하신 이유를 여쭤봐도 될까요?
`IllegalHttpRequestException`은 프로그램의 오류라기보다는 클라이언트(외부)의 의도치 않은 사용으로 인한 예외를 규정하기 위한 것이라 `Exception`을 상속하도록 하였습니다.
`RuntimeException`을 상속해도 문제는 없습니다만 개인적으로는 `Exception`을 상속하는 것이 의미가 더 맞다고 생각합니다.

johngrib replied:
일종의 컨벤션이라 둘 중 무엇을 써도 상관은 없지만 Exception이 checked / unchecked 예외를 둘다 포괄하는 개념인 반면, RuntimeException은 unchecked 예외를 표현하는 개념이기 때문입니다. 작업하신 `resolve` 메소드 본문을 보니 `IllegalHttpRequestException`를 throw하는 내용을 갖고 있는데 메소드 시그니처에 throws를 붙여주지 않으셨더라고요. 그렇다면 @steve7867 님이 의도하신 건 RuntimeException이라고 본 거죠.

johngrib replied:
이런 지금 보니 IllegalArgumentException을 잘못 본 거였군요 :-)

![image](https://user-images.githubusercontent.com/1855714/195987217-ba091aef-1b07-4832-a0c9-c53ae20b0dd1.png)

steve7867 replied:
제가 커밋을 뒤죽박죽으로 날린 탓도 있습니다😭

johngrib commented on app/src/main/java/com/codesoom/assignment/models/HttpRequest.java:
```diff
@@ -0,0 +1,57 @@
+import java.io.InputStreamReader;
+import java.util.stream.Collectors;
+
+public class HttpRequest {
+    private final HttpMethod httpMethod;
+    private final Path path;
+    private final String requestBody;
```
이 4줄의 구성이 보기 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/HttpRequest.java:
```diff
@@ -0,0 +1,57 @@
+    private final String requestBody;
+
+    public HttpRequest(HttpExchange exchange) throws IllegalHttpRequestException {
+        this.httpMethod = extractHttpMethod(exchange);
+        this.path = extractPath(exchange);
+        this.requestBody = extractRequestBody(exchange);
```
좋습니다. 코딩 연습을 위해 다른 방식으로도 해볼까요? 이 세 작업을 생성자에서 하지 않고 각 getter에서 수행하도록 바꿔보세요.

steve7867 replied:
인스턴스 필드의 초기화를 생성자 호출 시점에 하지 않고, getter를 호출하는 시점에 초기화하는 방식(지연 방식)을 말씀하시는 건가요?

steve7867 replied:
getter 호출 시점에 초기화가 이루어질 수 있도록 변경하였습니다.

steve7867 committed:
코드 Formatting 및 메소드명 변경

isNumber() -> isNumberFormat()

steve7867 committed:
Utility 클래스의 생성자를 private으로 변경

steve7867 committed:
HttpRequest의 instance field의 초기화 방식 변경

생성자 호출 시점에 초기화하지 않고 getter 호출 시점에 초기화(지연 방식)

steve7867 committed:
Task의 id 필드 재할당 금지 및 코드 간소화

id 필드에 final을 붙혀 재할당을 금지하고 관련 코드 리팩토링
@JsonProperty를 이용한 코드 간소화

steve7867 committed:
HttpRequestServiceResolver 추가 및 관련 코드 리팩토링, ServiceDispatcher JavaDoc 작성

steve7867 committed:
checksMissingPartExists() 제거

steve7867 committed:
ServiceDispatcher JavaDoc 수정

steve7867 committed:
Http Request Body의 Json 포맷 여부 검증 기능 추가 및 관련 코드 리팩토링

steve7867 committed:
handle() JavaDoc 추가

steve7867 committed:
HttpRequestValidator Exception 메시지 수정 및 checksPathValid 테스트 코드 추가

steve7867 committed:
HttpStatusCode getter 추가, ServiceDispatcher 리팩토링

writeAndFlushAndClose()을 sendResponse()로 변경하고 response header 설정 부분까지 sendResponse()로 포함

steve7867 commented:
이번에도 커밋 수가 많네요ㅠㅠ
작업하고 돌아서면 자꾸 아이디어가 떠올라서 다시 코딩하다 보니 이렇게 됐습니다.
죄송합니다😭

johngrib commented:
> 이번에도 커밋 수가 많네요ㅠㅠ 작업하고 돌아서면 자꾸 아이디어가 떠올라서 다시 코딩하다 보니 이렇게 됐습니다. 죄송합니다😭

괜찮습니다 :-)

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/HttpRequestValidatorTest.java:
```diff
@@ -0,0 +1,33 @@
+        for (String invalidPath : invalidPathArr) {
+            IllegalHttpRequestPathException exception = assertThrows(IllegalHttpRequestPathException.class
+                    , () -> HttpRequestValidator.checksPathValid(invalidPath));
+            System.out.println(exception.getMessage());
```
```suggestion
```

steve7867 replied:
커밋 전에 지운다는 걸 깜빡했습니다.
지우겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/HttpRequestValidatorTest.java:
```diff
@@ -0,0 +1,33 @@
+    @DisplayName("URI 유효성 테스트 - invalid path")
+    void checksPathInvalidTest() {
+
+        String[] invalidPathArr = {null, "tasks", "1111", "/1111", "/tasks/exercise", "tasks/", "tasks/1", "/tasks/exercises/1"};
```
```suggestion
        final String[] invalidPathArr = {null, "tasks", "1111", "/1111", "/tasks/exercise", "tasks/", "tasks/1", "/tasks/exercises/1"};
```

steve7867 replied:
final 붙히는 것도 놓쳤네요.
감사합니다~

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/HttpRequestValidatorTest.java:
```diff
@@ -0,0 +1,33 @@
+    @Test
+    @DisplayName("URI 유효성 테스트 - valid path")
+    void checksPathValidTest() {
+        String[] validPathArr = {"/tasks", "/tasks/1", "/tasks/31273612783", "/tasks/"};
```
```suggestion
        final String[] validPathArr = {"/tasks", "/tasks/1", "/tasks/31273612783", "/tasks/"};
```

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringValidator.java:
```diff
@@ -0,0 +1,17 @@
+    private StringValidator() {
+    }
+
+    public static boolean isNumberFormat(String s) {
+        for (char c : s.toCharArray()) {
+            if (c < '0' || c > '9') {
+                return false;
+            }
+        }
+
+        return true;
+    }
```
추가 과제로 0 으로 시작하는 경우도 invalid하게 판별하도록 작업해 봅시다.

steve7867 replied:
```java
public static boolean isNumberFormatValid(String s) {
    if (s.equals("0")) {
        return true;
    }

    if (s.startsWith("0")) {
        return false;
    }

    for (int i = 0; i < s.length(); i++) {
        final char c = s.charAt(i);
        if (c < '0' || c > '9') {
            return false;
        }
    }

    return true;
}
```

인자가 "0"인 경우에는 true를 반환하고 그 외의 '0'으로 시작하는 문자열은 false를 반환하게끔 구현했습니다.

johngrib replied:
좋습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/services/EditService.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.io.IOException;
+
+public class EditService implements HttpRequestService {
```
이름에 Task를 넣는 것도 괜찮겠습니다.
```suggestion
public class TaskEditService implements HttpRequestService {
```

steve7867 replied:
맞습니다.
수정하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,21 @@
+        this.title = title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
```
추가 과제로 이 Task가 업데이트될 때마다 업데이트된 시간을 기록하는 `updated` 필드도 추가해 보도록 합시다.

steve7867 replied:
```java
    private final LocalDateTime regDate;
    private LocalDateTime modDate;

    public Task(Long id, String title, LocalDateTime regDate) {
        this.id = id;
        this.title = title;
        this.regDate = regDate;
        this.modDate = regDate;
    }

    public void changeTitle(String title) {
        this.title = title;
        modDate = LocalDateTime.now();
    }
```

regDate(생성날짜), modDate(변경날짜) 필드를 추가하고, 객체를 생성할 때 regDate, modDate가 생성 시점의 시간으로 초기화되고, title을 변경할 때 modDate가 변경 시점의 시간으로 재할당되도록 구현했습니다.

johngrib replied:
좋습니다!

steve7867 committed:
콘솔 출력문 삭제

Co-authored-by: JohnGrib <johngrib82@gmail.com>

steve7867 committed:
invalidPathArr 재할당 금지

Co-authored-by: JohnGrib <johngrib82@gmail.com>

steve7867 committed:
validPathArr 재할당 금지

Co-authored-by: JohnGrib <johngrib82@gmail.com>

steve7867 committed:
EditService -> TaskEditService

Co-authored-by: JohnGrib <johngrib82@gmail.com>

steve7867 committed:
Task 등록날짜, 변경날짜 필드 추가 및 관련 코드 리팩토링

steve7867 committed:
isNumberFormat() 수정, JavaDoc 추가, 테스트 코드 추가

앞자리가 '0'으로 시작하는지 검증하는 로직 추가

steve7867 committed:
클래스명 변경으로 인한 파일명 변경 및 관련 코드 리팩토링

steve7867 committed:
Task 버그 수정, ServiceDispatcher 버그 수정

Task: 객체 생성 시 modDate가 초기화 되지 않는 버그 수정
ServiceDispatcher: http status code가 204(no content)일 때, response body의 길이를 0으로 설정

steve7867 committed:
메소드명 변경 누락 부분 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringValidator.java:
```diff
@@ -5,8 +5,25 @@ public final class StringValidator {
 
-    public static boolean isNumberFormat(String s) {
-        for (char c : s.toCharArray()) {
+    /**
+     * This method validates whether the given argument is suitable uri resource id.
+     * Suitable uri resource id means not including any other characters than digit characters ('0' ~ '9')
+     * and not starting with '0' except for "0" by itself.
+     *
+     * @param s a string to be validated
+     * @return true if given argument is suitable, false if not suitable
+     */
```
코드숨을 진행하는 동안 JavaDoc은 한국어로 작성해 보도록 합시다. 먼저 한국어로 잘 작성할 수 있게 되면 그 이후에 영어로 작성하는 연습을 따로 해보는 거죠.

steve7867 replied:
네 한국어로 잘 작성하는 것부터 연습하겠습니다.

johngrib reviewed (approved):
좋습니다.
